cli/tests/e2e.rs

```diff
@@ -0,0 +1,219 @@
+            config.clone(),
+        )
+        .await
+        .unwrap();
+        tokio::fs::write(format!("{dir}/.simperby/auth.json"), auth.clone())
+            .await
+            .unwrap();
+
+        run_command(format!(
+            "cd {dir} && git remote add server git://127.0.0.1:{port}/"
+        ))
+        .await;
+    }
+
+    // Add files for cli.
+    let config = Config { peers: Vec::new() };
+    let config = serde_spb::to_string(&config).unwrap();
+    let auth = Auth {
+        private_key: keys[3].1.clone(),
+    };
+    let auth = serde_spb::to_string(&auth).unwrap();
+    let server_config = serde_spb::to_string(&server_config).unwrap();
+    tokio::fs::write(
+        format!("{server_dir}/.simperby/config.json"),
+        config.clone(),
+    )
+    .await
+    .unwrap();
+    tokio::fs::write(format!("{server_dir}/.simperby/auth.json"), auth.clone())
+        .await
+        .unwrap();
+    tokio::fs::write(
+        format!("{server_dir}/.simperby/server_config.json"),
+        server_config.clone(),
+    )
+    .await
+    .unwrap();
+
+    // Start the server.
+    let mut child = tokio::process::Command::new(cli_path.clone())
+        .arg(server_dir)
+        .arg("serve")
+        .spawn()
+        .unwrap();
+    tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+
+    // Step 1: create an agenda and propagate it
+    log::info!("STEP 1");
+    run_command(format!("{cli_path} {} create agenda", clients_path[0])).await;
+    let raw_repo = RawRepository::open(clients_path[0].clone().as_str())
+        .await
+        .unwrap();
+    let branches = raw_repo
+        .list_branches()
+        .await
+        .unwrap()
+        .into_iter()
+        .filter(|b| b.starts_with("a-"))
+        .collect::<Vec<String>>();
+    let agenda_branch = branches.first().unwrap();
+
+    sync_each_other(cli_path.clone(), clients_path.clone()).await;
+    for path in clients_path.clone() {
+        run_command(format!("{cli_path} {path} vote {agenda_branch}")).await;
+    }
+    sync_each_other(cli_path.clone(), clients_path.clone()).await;
+
+    // Step 2: create block and run consensus
+    log::info!("STEP 2");
+
+    run_command(format!("{cli_path} {} create block", clients_path[0])).await;
+    sync_each_other(cli_path.clone(), clients_path.clone()).await;
+    for path in clients_path.clone() {
+        run_command(format!("{cli_path} {path} consensus")).await;
+    }
+    sync_each_other(cli_path.clone(), clients_path.clone()).await;
+    for path in clients_path.clone() {
+        run_command(format!("{cli_path} {path} consensus")).await;
+    }
+    sync_each_other(cli_path.clone(), clients_path.clone()).await;
+    for path in clients_path.clone() {
+        run_command(format!("{cli_path} {path} consensus")).await;
+    }
+    sync_each_other(cli_path.clone(), clients_path.clone()).await;
+    for path in clients_path.clone() {
+        run_command(format!("{cli_path} {path} consensus")).await;
```

Comment:
Do you know why we call consensus command four times? I thought we only need two.

Reply:
I show you the results of client's `progress_for_order` in the order of client1(proposer), client2, and client3.
1. [Proposed, NonNilPreVoted], [], []
2. [], [NonNilPreVoted], [NonNilPreVoted]
3. [NonNilPreCommitted], [NonNilPreCommitted], [NonNilPreCommitted]
4. [Finalized], [Finalized], [Finalized]

These states are defined in `consensus/src/lib.rs` `ProgressResult` and that's why we need consensus commands four times.

Reply:
Understood.